### PR TITLE
Avoid duplicated deployment of kar on L1 guest

### PR DESCRIPTION
**Pull request by qcheng-redhat**
ID: 1814562

To avoid duplicated deployment of kar on L1 guest. 

Signed-off-by: Qinghua Cheng <qcheng@redhat.com>

### CODE REVIEW

**Comment by qcheng-redhat** [comment]
@coli @luckyh Please help review. Thanks.

**Comment by qcheng-redhat** [comment]
Add @yfu, please help to review. Thanks.

**Comment by yanan-fu** [comment]
@qcheng-redhat This patch assume `kar_folder.stat.exists` also means the pre bootstrap passed. This is a risky.
One idea in mind is to save the bootstrap status in a special file in vm, check this flag together with `kar_folder.stat.exists`, if it is false, clean env (remove kar, uninstall avocado, avocado-vt, aexpect) and redeploy. @luckyh do you have any other suggestion about this ? thanks!

**Comment by qcheng-redhat** [comment]
Add uninstall avocado, avocado-vt and aexpect in task clean env when bootstrap failed.

File: qemu/deps/playbook.yml [hunk 1]
```diff
@@ -40,18 +39,34 @@
...
         chdir: '{{kar_local}}'
+      when: not kar_folder.stat.exists
       register: bootstrap_result
+      ignore_errors: True
+
+    - name: Clean env when bootstrap failed
+      shell: |
+        rm -rf '{{kar_local}}'
+        pip uninstall -y `pip list | grep avocado`
+        pip uninstall -y `pip list | grep aexpect`
```
Comment (yanan-fu) [inline qemu/deps/playbook.yml]: If these package still have not be installed before get `bootstrap_result` failed, will get error: `ERROR: You must give at least one requirement to uninstall`, suggest use the 'pip' module for ansible:
```
      - name: clean env ****
        pip:
          name: aexpect, avocado-framework, avocado-framework-plugin-vt
          state: absent
```

Comment (qcheng-redhat) [reply qemu/deps/playbook.yml]: If use pip, we need to remove kar directory in another task. This is not expected.

Comment (yanan-fu) [reply qemu/deps/playbook.yml]: > If use pip, we need to remove kar directory in another task. This is not expected.

Split to two task is ok, it is not must to put them in one task i think.  Just clean the env is ok.

**Comment by qcheng-redhat** [comment]
Split remove kar folder and uninstall pkgs into two tasks.

File: qemu/deps/playbook.yml [hunk 2]
```diff
@@ -40,18 +39,38 @@
...
       command: ./Bootstrap.sh {{bs_options}}
       args:
         chdir: '{{kar_local}}'
+      when: not kar_folder.stat.exists
       register: bootstrap_result
+      ignore_errors: True
+
+    - name: Remove kar folder
+      shell:
+        rm -rf '{{kar_local}}'
```
Comment (yanan-fu) [inline qemu/deps/playbook.yml]: prefer to use the ansible module `file`:
```
- name: Remove kar folder when bootstrap failed
  file:
    path:  '{{kar_local}}'
    state: absent
```

Comment (qcheng-redhat) [reply qemu/deps/playbook.yml]: Fixed.

File: qemu/deps/playbook.yml [hunk 3]
```diff
@@ -40,18 +39,39 @@
...
 
     - name: Run bootstrap script
       command: ./Bootstrap.sh {{bs_options}}
       args:
         chdir: '{{kar_local}}'
+      when: not kar_folder.stat.exists
       register: bootstrap_result
+      ignore_errors: True
+
+    - name: Remove kar folder
```
Comment (yanan-fu) [inline qemu/deps/playbook.yml:62]: `Remove kar folder when bootstrap failed `  ?

Comment (qcheng-redhat) [reply qemu/deps/playbook.yml:62]: Yes, if bootstrap failed, remove kar folder and uninstall rpms.

Comment (yanan-fu) [reply qemu/deps/playbook.yml:62]: @qcheng-redhat I mean update the comment for this task, to make it more clear.

Comment (qcheng-redhat) [reply qemu/deps/playbook.yml]: Updated task name.

**Comment by yanan-fu** [comment]
LGTM
@luckyh  Do you have any comments ? Thanks!

**Comment by qcheng-redhat** [comment]
@PaulYuuu please help review. Thanks

File: qemu/deps/playbook.yml [hunk 4]
```diff
@@ -40,18 +39,39 @@
...
+        path: '{{kar_local}}'
+        state: absent
+      when: bootstrap_result is failed
+
+    - name: Clean env when bootstrap failed
+      pip:
+        name: aexpect, avocado-framework, avocado-framework-plugin-vt, avocado-framework-plugin-result-html
+        state: absent
+      when: bootstrap_result is failed
+      failed_when:
```
Comment (PaulYuuu) [inline qemu/deps/playbook.yml:73]: If add `failed_when` here, it's not clear when the trigger is wrong, like:
```
TASK [Clean env when bootstrap failed]
...
failed
...
```
Suggestions:
```
fail:
  msg: xxx
when:
```
or
```
asset:
  that: bootstrap_result is changed
  success_msg: bootstrap complete
  fail_msg: bootstrap failed

File: qemu/deps/playbook.yml [hunk 5]
```diff
@@ -40,18 +39,39 @@
...
+      stat:
+         path: '{{kar_local}}'
+      register: kar_folder
+
+    - name: Clone kar
       git:
         repo: '{{kar_repo}}'
         dest: '{{kar_local}}'
         force: yes
+      when: not kar_folder.stat.exists
```
[the remark this answers is not in the document]
Comment (qcheng-redhat) [reply qemu/deps/playbook.yml:52]: Only if not kar_folder.stat.exists is true, this task will be executed.

File: qemu/deps/playbook.yml [hunk 6]
```diff
@@ -40,18 +39,39 @@
...
 
     - name: Run bootstrap script
       command: ./Bootstrap.sh {{bs_options}}
       args:
         chdir: '{{kar_local}}'
+      when: not kar_folder.stat.exists
```
[the remark this answers is not in the document]
Comment (qcheng-redhat) [reply qemu/deps/playbook.yml:58]: Could you explain a bit more for this comment ? Thanks.

File: qemu/deps/playbook.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (yanan-fu) [reply qemu/deps/playbook.yml:52]: Good catch ~

[the remark this answers is not in the document]
Comment (PaulYuuu) [reply qemu/deps/playbook.yml:52]: > Only if not kar_folder.stat.exists is true, this task will be executed.

Yes, so I think we can remove `force`.

File: qemu/deps/playbook.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (PaulYuuu) [reply qemu/deps/playbook.yml:58]: Ah, my fault, it's correct.

File: qemu/deps/playbook.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (qcheng-redhat) [reply qemu/deps/playbook.yml:51]: Fixed.

File: qemu/deps/playbook.yml (hunk 4 above)
Comment (qcheng-redhat) [reply qemu/deps/playbook.yml:72]: Plan to resolve this comment in the next patch. Is it ok?

Comment (qcheng-redhat) [reply qemu/deps/playbook.yml]: Changed failed_when to fail module with failed msg.

**Comment by yanan-fu** [comment]
Thanks all!

**Pull request merged**